examples/keyset.rs

```diff
@@ -0,0 +1,524 @@
+    report_actions(actions, &ks);
+
+    save_keyset(filename, &ks);
+}
+
+fn do_propagation2(filename: &str, args: &[String]) {
+    if args.len() != 2 {
+        eprintln!("Usage: ... propagation2-complete <roll-type> <ttl>");
+        exit(1);
+    }
+    let rolltype = &args[0];
+    let ttl = &args[1];
+
+    let rolltype = str_to_rolltype(rolltype);
+    let ttl = ttl.parse().unwrap();
+
+    let mut ks = load_keyset(filename);
+
+    let actions = ks.propagation2_complete(rolltype, ttl);
+    report_actions(actions, &ks);
+
+    save_keyset(filename, &ks);
+}
+
+fn do_cache_expired2(filename: &str, args: &[String]) {
+    if args.len() != 1 {
+        eprintln!("Usage: ... cache-expired2 <roll-type>");
+        exit(1);
+    }
+    let rolltype = &args[0];
+
+    let rolltype = str_to_rolltype(rolltype);
+
+    let mut ks = load_keyset(filename);
+
+    let actions = ks.cache_expired2(rolltype);
+    report_actions(actions, &ks);
+
+    save_keyset(filename, &ks);
+}
+
+fn do_done(filename: &str, args: &[String]) {
+    if args.len() != 1 {
+        eprintln!("Usage: ... done <roll-type>");
+        exit(1);
+    }
+    let rolltype = &args[0];
+
+    let rolltype = str_to_rolltype(rolltype);
+
+    let mut ks = load_keyset(filename);
+
+    let actions = ks.roll_done(rolltype);
+    report_actions(actions, &ks);
+
+    save_keyset(filename, &ks);
+}
+
+fn do_actions(filename: &str, args: &[String]) {
+    if args.len() != 1 {
+        eprintln!("Usage: ... actions <roll-type>");
+        exit(1);
+    }
+    let rolltype = &args[0];
+
+    let rolltype = str_to_rolltype(rolltype);
+
+    let mut ks = load_keyset(filename);
+
+    let actions = ks.actions(rolltype);
+    report_actions(actions, &ks);
+}
+
+fn do_status(filename: &str, args: &[String]) {
+    if !args.is_empty() {
+        eprintln!("Usage: ... status");
+        exit(1);
+    }
+
+    let ks = load_keyset(filename);
+
+    println!("Keys:");
+    let keys = ks.keys();
+    for key in keys {
+        match key.keytype() {
+            KeyType::Ksk(keystate)
+            | KeyType::Zsk(keystate)
+            | KeyType::Include(keystate) => {
+                println!(
+                    "\t{} {}",
+                    key.pubref(),
+                    key.privref().unwrap_or_default(),
+                );
+                println!("\t\tState: {}", keystate);
+            }
+            KeyType::Csk(keystate_ksk, keystate_zsk) => {
+                println!(
+                    "\t{} {}",
+                    key.pubref(),
+                    key.privref().unwrap_or_default(),
+                );
+                println!("\t\tKSK role state: {}", keystate_ksk,);
+                println!("\t\tZSK role state: {}", keystate_zsk,);
+            }
+        }
+        let ts = key.timestamps();
+        println!(
+            "\t\tCreated: {}",
+            ts.creation()
+                .map_or("<empty>".to_string(), |x| x.to_string()),
```

Comment:
`format_args!()` is the right tool here, but it's really unfortunate that it's harder to use than `.to_string()`.

Reply:
How would that work?

Reply:
`ts.creation().map_or(format_args!("<empty>"), |x| format_args!("{}", x))`.  But it's longer and more painful to use, so I suggest sticking with `.to_string()`.